docs/tutorials/tutorial-2.md

```diff
@@ -0,0 +1,794 @@
+```
+for this line, which uses `.` instead of `~`:
+```c
+accounts.udict_set_builder(256, sender, begin_cell().store_coins(balance));
+```
+and save the file as `deposit_bug_bank.fc`.
+
+The change has the effect that the `accounts` dictionary does not get updated. This means that if a deposit message arrives for an account id that initially does not exist in the dictionary, then after the message gets processed, the account would still not exist in the dictionary. 
+
+Indeed, if we run TSA again on the modified source code:
+
+```
+java -jar tsa.jar custom-checker --checker checker_1a.fc --contract func deposit_bug_bank.fc --fift-std imports/lib -o report_bug_1a.sarif
+```
+
+We should see a non-empty `"results"` list in `report_bug_1a.sarif`, with a `"message"` property:
+
+```
+"message": {
+    "text": "TvmFailure(exit=TVM user defined error with exit code 257, phase=COMPUTE_PHASE)"
+}
+```
+that indicates that TSA found an instantiation that threw the exception at the end of `checker_1a.fc`, i.e., TSA found a violation to invariant (1.a).
+
+Looking further down in `report_bug_1a.sarif`, at the `"fetchedValues"` property, we should see the variables that were inspected with function `tsa_fetch_value` inside the checker (comments added for convenience):
+```
+"fetchedValues": {
+    "1": "73786976294838206464",     [Deposit amount]
+    "2": "0",                        [Balance after deposit message]
+    "3": "0",                        [Account id]
+    "4": "0"                         [Found flag after deposit message]
+}
+```
+
+We see that, even though the deposit amount is higher than the fees (`10000000`), the account id still does not exist in the dictionary, because the found flag after the deposit message is `0`, i.e., `false`.
+
+In case that TSA is still not able to find violations after manually injecting bugs in the source code, it is recommended to do the following sanity check. 
+
+It may be the case that, in the checker, the constraints between the functions `tsa_forbid_failures` and `tsa_allow_failures` are contradictory or always throw exceptions. If the constraints are contradictory or always fail, TSA will not be able to find an instantiation that reaches `tsa_allow_failures`, because it is impossible, producing an empty report as a result. 
+
+Therefore, it is important to sanity check your constraints by writing a checker that always throws an exception immediately after the call to function `tsa_allow_failures`, like so:
+```c
+() main() impure {
+    tsa_forbid_failures();
+
+    ;; All your constraints
+
+    tsa_allow_failures();
+
+    throw(1000);  ;; Unconditionally throw an exception
+}
+```
+If TSA does not report the exception `1000` when running the above checker, then there is something wrong with your constraints, and you need to revise them: they may be contradictory or always throw an exception.
+
+Having explored the results for invariant (1.a), let us now switch to the report for invariant (2.a). 
```

Comment:
Maybe add subtitles for running (or just interpreting the results) the checker for (1.a) and for (2.a)?